### ISSUE TITLE
Bad suggestion `dyn pub` on proc macro (#61963 regressed)

### ISSUE THREAD

**Issue by SNCPlay42** [rust-lang/rust#74043]
#72306 causes the compiler to emit this bad suggestion in [the test](https://github.com/rust-lang/rust/blob/master/src/test/ui/suggestions/issue-61963.stderr) for #61963:

```
error: trait objects without an explicit `dyn` are deprecated
  --> $DIR/issue-61963.rs:18:1
   |
LL | pub struct Foo {
   | ^^^ help: use `dyn`: `dyn pub`
```

This is the behaviour the test is supposed to guard against (almost - the original suggestion was to insert `dyn` before the proc macro attribute), so the issue is regressed.

The issue was originally fixed nearly a year ago, so this is a regression from stable.

**Comment by LeSeulArtichaut** [comment]
cc @Aaron1011 @davidtwco

**Comment by Aaron1011** [comment]
I don't think this is actually an issue anymore. The original issue was caused by the compiler losing span information, causing tokens emitted by a proc-macro to end up with a span (the proc-macro attribute) that the author did not intend.

However, the test is explicitly setting the span of the token in question:

https://github.com/rust-lang/rust/blob/0cd7ff7ddfb75a38dca81ad3e76b1e984129e939/src/test/ui/suggestions/auxiliary/issue-61963-1.rs#L13-L15

However, the first token in the stream no longer has a dummy span, so this proc-macro ends up explicitly spanning the trait object token with the span of the `pub` token.

I don't think this is a bug - when a proc-macro explicitly sets the span of a token, I don't think the compiler should be trying to second-guess that decision. There are legitimate use cases that involve causing an error to be displayed to be displayed in a unusual location - e.g. pointing to a field definition for a trait bound that the field type fails to meet.

We might want to alter the test documentation to better explain its current behavior (while still guarding for regressions to the original behavior).

**Comment by SNCPlay42** [comment]
So this is different from the original situation in #61963? There was discussion there about whether the problem was a bug in the proc macros or the compiler.

> There are legitimate use cases that involve causing an error to be displayed to be displayed in a unusual location - e.g. pointing to a field definition for a trait bound that the field type fails to meet.

This I understand, but should the compiler then emit suggestions that produce invalid syntax? They at least shouldn't be reported as `MachineApplicable`. This causes `cargo fix` to try to automatically apply the fix, and then fail to compile the result, which it says "likely indicates a bug in either rustc or cargo itself".

**Comment by Aaron1011** [comment]
> So this is different from the original situation in #61963? There was discussion there about whether the problem was a bug in the proc macros or the compiler.

Yes - there is no longer any implicit respanning occuring, which was the root of the original issue.

> This I understand, but should the compiler then emit suggestions that produce invalid syntax? 

I think this is unsolvable in the general case - the span could point to a token  `Foo` instead of `Pub`, which could either be a trait `Foo` or a struct `Foo`.

However, I think it would be fine to add in some special cases (e.g. only suggest `dyn <ident>`, not `dyn <keyword>` if people are hitting this in real code.

**Comment by Dylan-DPC** [comment]
Marking it as p-medium based on the [discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2374043.20Bad.20suggestion.20.60dyn.20pub.60.20on.20proc.20macro.20)
